File: 04_quicksort/php/05_quicksort.php
<?php

function quicksort(array $array) {
    if (count($array) < 2) {
        // base case, arrays with 0 or 1 element are already "sorted"
        return $array;
    } else {
        // recursive case
        $pivot = $array[0];
        var_dump($array);

        // sub-array of all the elements less than the pivot
        $less = array_filter(array_slice($array, 1), function($el) use ($pivot) { return $el <= $pivot; });

        // sub-array of all the elements greater than the pivot
        $greater = array_filter(array_slice($array, 1), function($el) use ($pivot) { return $el > $pivot; });
        return array_merge(quicksort($less), [$pivot], quicksort($greater));
  }
}

var_dump(quicksort([10, 5, 2, 3])); // [2, 3, 5, 10]

File: 04_quicksort/golang/05_quicksort.go
package main

import "fmt"

func quicksort(list []int) []int {
	if len(list) < 2 {
		return list
	} else {
		pivot := list[0]

		var less = []int{}
		var greater = []int{}
		for _, num := range list[1:] {
			if pivot > num {
				less = append(less, num)
			} else {
				greater = append(greater, num)
			}
		}

		less = append(quicksort(less), pivot)
		greater = quicksort(greater)
		return append(less, greater...)
	}
}

func main() {
	fmt.Println(quicksort([]int{10, 5, 2, 3}))
}
